### PR TITLE
JIT: Improve weight distribution in internal blocks produced by casts

### LINKED ISSUE #50030: [JIT][PGO] Incorrect weights produced by expand-casts optimizations
```csharp
static string Foo(object o) => (string)o;
```
Currently produces:
```asm
; Method Program:Foo(System.Object):System.String
G_M28009_IG01:              ;; offset=0000H
       4883EC28             sub      rsp, 40
						;; bbWeight=1    PerfScore 0.25

G_M28009_IG02:              ;; offset=0004H
       48894C2430           mov      gword ptr [rsp+30H], rcx
       488BC1               mov      rax, rcx
       4885C0               test     rax, rax
       740F                 je       SHORT G_M28009_IG04
						;; bbWeight=1    PerfScore 2.50

G_M28009_IG03:              ;; offset=0011H
       48BAB873D0E9F97F0000 mov      rdx, 0x7FF9E9D073B8
       483910               cmp      qword ptr [rax], rdx
       7505                 jne      SHORT G_M28009_IG05
						;; bbWeight=0.25 PerfScore 0.81

G_M28009_IG04:              ;; offset=0020H
       4883C428             add      rsp, 40
       C3                   ret      
						;; bbWeight=1    PerfScore 1.25

G_M28009_IG05:              ;; offset=0025H
       488BCA               mov      rcx, rdx
       488B542430           mov      rdx, gword ptr [rsp+30H]
       E83E7BACFF           call     CORINFO_HELP_CHKCASTCLASS_SPECIAL
       CC                   int3     
						;; bbWeight=0    PerfScore 0.00
; Total bytes of code: 51
```

Note that 0.25 weight which looks incorrect.

Same for some other cast patterns.

category:correctness
theme:profile-feedback
skill-level:intermediate
cost:small
impact:small

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes https://github.com/dotnet/runtime/issues/50030

```csharp
static bool Test(object x) => x is string;
```
Codegen diff: https://www.diffchecker.com/MEyYf7UH
So we just let `optSetBlockWeights` to handle weights.

jit-diff:
```
PMI CodeSize Diffs for System.Private.CoreLib.dll, framework assemblies for  default jit

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 52984741
Total bytes of diff: 52983438
Total bytes of delta: -1303 (-0.00% of base)
    diff is an improvement.


Top file regressions (bytes):
          87 : System.Security.AccessControl.dasm (0.10% of base)
          82 : FSharp.Core.dasm (0.00% of base)
          45 : System.Data.Common.dasm (0.00% of base)
          24 : Newtonsoft.Json.Bson.dasm (0.03% of base)
          19 : System.Private.DataContractSerialization.dasm (0.00% of base)
          13 : System.IO.Compression.ZipFile.dasm (0.23% of base)
          13 : System.DirectoryServices.Protocols.dasm (0.01% of base)
           8 : System.Linq.Queryable.dasm (0.00% of base)
           7 : System.Private.Xml.Linq.dasm (0.00% of base)
           7 : System.Diagnostics.DiagnosticSource.dasm (0.01% of base)
           6 : System.Runtime.Caching.dasm (0.01% of base)
           6 : System.Data.OleDb.dasm (0.00% of base)
           6 : System.Net.Security.dasm (0.00% of base)
           5 : System.ObjectModel.dasm (0.00% of base)
           3 : System.Text.Json.dasm (0.00% of base)
           3 : System.ComponentModel.Composition.dasm (0.00% of base)
           1 : System.Formats.Asn1.dasm (0.00% of base)
           1 : System.IO.FileSystem.Watcher.dasm (0.01% of base)

Top file improvements (bytes):
        -465 : System.Memory.dasm (-0.17% of base)
        -399 : Microsoft.CodeAnalysis.VisualBasic.dasm (-0.01% of base)
        -106 : System.Private.CoreLib.dasm (-0.00% of base)
        -102 : System.Private.Xml.dasm (-0.00% of base)
        -101 : System.ComponentModel.Composition.Registration.dasm (-0.11% of base)
         -76 : System.Configuration.ConfigurationManager.dasm (-0.02% of base)
         -61 : System.Composition.Convention.dasm (-0.06% of base)
         -59 : Microsoft.CodeAnalysis.CSharp.dasm (-0.00% of base)
         -57 : System.DirectoryServices.dasm (-0.01% of base)
         -47 : System.DirectoryServices.AccountManagement.dasm (-0.01% of base)
         -42 : System.Linq.Expressions.dasm (-0.01% of base)
         -30 : System.CodeDom.dasm (-0.02% of base)
         -20 : System.Drawing.Common.dasm (-0.01% of base)
         -19 : Microsoft.Diagnostics.Tracing.TraceEvent.dasm (-0.00% of base)
         -12 : System.Speech.dasm (-0.00% of base)
         -11 : Microsoft.CSharp.dasm (-0.00% of base)
         -10 : Newtonsoft.Json.dasm (-0.00% of base)
          -6 : System.Data.Odbc.dasm (-0.00% of base)
          -6 : System.Runtime.Serialization.Formatters.dasm (-0.01% of base)
          -6 : System.Security.Cryptography.Xml.dasm (-0.00% of base)

40 total files with Code Size differences (22 improved, 18 regressed), 231 unchanged.

Top method regressions (bytes):
         116 ( 2.80% of base) : FSharp.Core.dasm - FSharpChoice`2:CompareTo(Object,IComparer):int:this (8 methods)
          78 ( 1.80% of base) : System.Security.AccessControl.dasm - CommonAcl:RemoveQualifiedAces(SecurityIdentifier,int,int,ubyte,bool,int,Guid,Guid):bool:this
          75 ( 0.40% of base) : System.Data.Common.dasm - BinaryNode:EvalBinaryOp(int,ExpressionNode,ExpressionNode,DataRow,int,ref):Object:this
          58 ( 0.86% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:ReportOverloadResolutionFailureAndProduceBoundNode(VisualBasicSyntaxNode,int,ArrayBuilder`1,ImmutableArray`1,TypeSymbol,ImmutableArray`1,ImmutableArray`1,DiagnosticBag,VisualBasicSyntaxNode,BoundMethodOrPropertyGroup,Symbol,bool,BoundTypeExpression,Symbol,Location):BoundExpression:this
          44 ( 1.37% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - SourceNonPropertyAccessorMethodSymbol:GetReturnType(SourceModuleSymbol,byref,DiagnosticBag):TypeSymbol:this
          40 ( 1.85% of base) : System.Composition.Convention.dasm - MethodExpressionAdapter:SelectMethods(Expression`1):MethodInfo (8 methods)
          38 ( 2.47% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Binder:ProcessParameterlessCrefMemberLookupResults(ImmutableArray`1,int,MemberCrefSyntax,TypeArgumentListSyntax,byref,DiagnosticBag):ImmutableArray`1:this
          35 ( 2.88% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - SyntaxTreeSemanticModel:GetTypeOrNamespaceSymbolInfoNotInMember(TypeSyntax,int):SymbolInfo:this
          26 ( 5.50% of base) : Microsoft.CodeAnalysis.CSharp.dasm - MatchSymbols:AreTypesEqual(TypeSymbol,TypeSymbol):bool:this
          24 ( 0.88% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:ComputeVariableType(LocalSymbol,ModifiedIdentifierSyntax,AsClauseSyntax,EqualsValueSyntax,byref,byref,DiagnosticBag):TypeSymbol:this
          24 ( 4.78% of base) : Microsoft.CodeAnalysis.CSharp.dasm - CodeGenerator:CanUseCallOnRefTypeReceiver(BoundExpression):bool:this
          23 ( 2.10% of base) : Newtonsoft.Json.dasm - BsonBinaryWriter:CalculateSize(BsonToken):int:this
          23 ( 2.10% of base) : Newtonsoft.Json.Bson.dasm - BsonBinaryWriter:CalculateSize(BsonToken):int:this
          22 ( 5.16% of base) : System.Private.Xml.dasm - XmlQueryRuntime:MatchesXmlType(XPathItem,int):bool:this (2 methods)
          20 ( 0.67% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - StateMachineMethodToClassRewriter:PossibleStateMachineScope(ImmutableArray`1,BoundNode):BoundNode:this (8 methods)
          18 ( 3.93% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - PEDeltaAssemblyBuilder:GetAnonymousTypeKey(NamedTypeSymbol):AnonymousTypeKey
          18 ( 0.53% of base) : System.ComponentModel.TypeConverter.dasm - ColorConverter:ConvertTo(ITypeDescriptorContext,CultureInfo,Object,Type):Object:this
          16 ( 4.65% of base) : System.Private.Xml.dasm - XmlSerializationReader:UnknownNode(XmlNode,Object,String):this
          16 ( 3.89% of base) : System.Private.Xml.Linq.dasm - XDocument:ValidateNode(XNode,XNode):this
          16 ( 2.18% of base) : Microsoft.CSharp.dasm - SymbolLoader:HasImplicitReferenceConversion(CType,CType):bool

Top method improvements (bytes):
        -444 (-8.83% of base) : System.Memory.dasm - SequenceReader`1:GetNextSpan():this (6 methods)
        -220 (-3.29% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:BindObjectCreationExpression(VisualBasicSyntaxNode,ArgumentListSyntax,TypeSymbol,VisualBasicSyntaxNode,ImmutableArray`1,ImmutableArray`1,BoundObjectInitializerExpressionBase,DiagnosticBag,VisualBasicSyntaxNode):BoundExpression:this
        -101 (-1.46% of base) : System.Composition.Convention.dasm - ConstructorExpressionAdapter:ParseSelectConstructor(Expression`1):this (8 methods)
        -101 (-1.46% of base) : System.ComponentModel.Composition.Registration.dasm - ConstructorExpressionAdapter:ParseSelectConstructor(Expression`1):this (8 methods)
         -68 (-4.21% of base) : System.Private.CoreLib.dasm - ModuleBuilder:GetMethodTokenNoLock(MethodInfo,bool):int:this
         -63 (-1.98% of base) : System.Configuration.ConfigurationManager.dasm - MgmtConfigurationRecord:SaveAs(String,int,bool):this
         -61 (-2.87% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:CheckViability(Symbol,int,int,TypeSymbol,byref):SingleLookupResult:this
         -50 (-2.53% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - SynthesizedEventAccessorSymbol:ConstructFieldLikeEventAccessorBody_WinRT(SourceEventSymbol,bool,VisualBasicCompilation,DiagnosticBag):BoundBlock
         -49 (-3.22% of base) : System.DirectoryServices.AccountManagement.dasm - StoreCtx:BuildFilterSet(Principal,ref,QbeFilterDescription):this
         -46 (-2.69% of base) : Microsoft.CodeAnalysis.CSharp.dasm - ExpressionLambdaRewriter:VisitInitializer(BoundExpression,byref):BoundExpression:this
         -34 (-1.35% of base) : System.Linq.Expressions.dasm - LightCompiler:CompileAddress(Expression,int):ByRefUpdater:this
         -34 (-2.12% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceAttributeData:GetTargetAttributeSignatureIndex(Symbol,AttributeDescription):int:this
         -32 (-2.08% of base) : Microsoft.CSharp.dasm - ComTypeDesc:GetMemberNames(bool):ref:this
         -29 (-1.44% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:BindIndexedInvocationExpression(InvocationExpressionSyntax,BoundExpression,ImmutableArray`1,ImmutableArray`1,ImmutableArray`1,bool,byref,DiagnosticBag):BoundExpression:this
         -27 (-1.74% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Binder:GetRewrittenAttributeConstructorArguments(byref,MethodSymbol,ImmutableArray`1,ImmutableArray`1,AttributeSyntax,DiagnosticBag,byref):ImmutableArray`1:this
         -27 (-0.71% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SourceMemberContainerTypeSymbol:ComputeInterfaceImplementations(DiagnosticBag,CancellationToken):ImmutableArray`1:this
         -26 (-0.73% of base) : System.Private.CoreLib.dasm - EventProvider:WriteEvent(byref,long,long,long,ref):bool:this
         -25 (-1.88% of base) : System.DirectoryServices.dasm - Domain:GetChildDomains():ArrayList:this
         -25 (-2.06% of base) : System.DirectoryServices.dasm - Forest:GetDomains():ArrayList:this
         -23 (-3.65% of base) : Microsoft.CodeAnalysis.CSharp.dasm - InitializerSemanticModel:BindEqualsValue(Binder,EqualsValueClauseSyntax,DiagnosticBag):BoundNode:this

Top method regressions (percentages):
          15 ( 6.41% of base) : Microsoft.CodeAnalysis.CSharp.dasm - ClsComplianceChecker:IsCompliantType(TypeSymbol,NamedTypeSymbol):bool:this
          26 ( 5.50% of base) : Microsoft.CodeAnalysis.CSharp.dasm - MatchSymbols:AreTypesEqual(TypeSymbol,TypeSymbol):bool:this
          22 ( 5.16% of base) : System.Private.Xml.dasm - XmlQueryRuntime:MatchesXmlType(XPathItem,int):bool:this (2 methods)
          12 ( 4.92% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - MemberLookup:LookupDefaultProperty(LookupResult,TypeSymbol,Binder,byref)
          24 ( 4.78% of base) : Microsoft.CodeAnalysis.CSharp.dasm - CodeGenerator:CanUseCallOnRefTypeReceiver(BoundExpression):bool:this
          16 ( 4.65% of base) : System.Private.Xml.dasm - XmlSerializationReader:UnknownNode(XmlNode,Object,String):this
          13 ( 4.36% of base) : Microsoft.CodeAnalysis.CSharp.dasm - LocalRewriter:CanChangeValueBetweenReads(BoundExpression,bool):bool
           6 ( 4.29% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - MemberLookup:AddLookupSymbolsInfo(LookupSymbolsInfo,NamespaceOrTypeSymbol,int,Binder)
           6 ( 3.95% of base) : Microsoft.CodeAnalysis.CSharp.dasm - Binder:AddMemberLookupSymbolsInfo(LookupSymbolsInfo,NamespaceOrTypeSymbol,int,Binder):this
          18 ( 3.93% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - PEDeltaAssemblyBuilder:GetAnonymousTypeKey(NamedTypeSymbol):AnonymousTypeKey
          16 ( 3.89% of base) : System.Private.Xml.Linq.dasm - XDocument:ValidateNode(XNode,XNode):this
          15 ( 3.87% of base) : Microsoft.CodeAnalysis.CSharp.dasm - CodeGenerator:EmitCondExpr(BoundExpression,bool):this
           8 ( 3.79% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - XmlWhitespaceChecker:VisitXmlElementEndTag(XmlElementEndTagSyntax):VisualBasicSyntaxNode:this
          11 ( 3.19% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - AnonymousTypeManager:CheckAndReportMissingSymbols(ArrayBuilder`1,DiagnosticBag):bool:this
           5 ( 3.12% of base) : Microsoft.CodeAnalysis.CSharp.dasm - SpecialMembersSignatureComparer:GetGenericTypeArgument(TypeSymbol,int):TypeSymbol:this
           6 ( 3.05% of base) : System.Net.Security.dasm - SecureChannel:MakeEx(X509Certificate):X509Certificate2
           8 ( 3.03% of base) : Microsoft.CodeAnalysis.CSharp.dasm - CSharpSemanticModel:GetMemberGroupFromNode(SyntaxNode,CancellationToken):ImmutableArray`1:this
          35 ( 2.88% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - SyntaxTreeSemanticModel:GetTypeOrNamespaceSymbolInfoNotInMember(TypeSyntax,int):SymbolInfo:this
         116 ( 2.80% of base) : FSharp.Core.dasm - FSharpChoice`2:CompareTo(Object,IComparer):int:this (8 methods)
           5 ( 2.59% of base) : Microsoft.CodeAnalysis.CSharp.dasm - BoundTreeRewriter:VisitConversion(BoundConversion):BoundNode:this

Top method improvements (percentages):
         -15 (-12.40% of base) : System.CodeDom.dasm - CodeValidator:ValidateCodeCompileUnit(CodeCompileUnit):this
        -444 (-8.83% of base) : System.Memory.dasm - SequenceReader`1:GetNextSpan():this (6 methods)
          -7 (-7.07% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - SymbolExtensions:ToErrorMessageArgument(Symbol):Object
          -9 (-6.12% of base) : System.Private.Xml.Linq.dasm - XAttribute:get_PreviousAttribute():XAttribute:this
          -6 (-5.56% of base) : Microsoft.CodeAnalysis.CSharp.dasm - PEModuleBuilder:ContainerIsGeneric(Symbol):bool
          -6 (-5.22% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - PEModuleBuilder:ContainerIsGeneric(Symbol):bool
         -10 (-5.18% of base) : System.Linq.Expressions.dasm - ChildRewriter:ShouldSaveToTemp(Expression):bool
          -7 (-4.76% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Symbol:get_CanBeReferencedByNameIgnoringIllegalCharacters():bool:this
          -9 (-4.76% of base) : Microsoft.CodeAnalysis.CSharp.dasm - ConversionsBase:HasImplicitBoxingTypeParameterConversion(TypeParameterSymbol,TypeSymbol,byref):bool:this
          -6 (-4.76% of base) : System.Data.Common.dasm - XPathNodePointer:GetRoot(XmlNode,byref):XmlNode
         -20 (-4.38% of base) : System.Drawing.Common.dasm - IconConverter:ConvertTo(ITypeDescriptorContext,CultureInfo,Object,Type):Object:this
         -68 (-4.21% of base) : System.Private.CoreLib.dasm - ModuleBuilder:GetMethodTokenNoLock(MethodInfo,bool):int:this
         -18 (-3.71% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - TypeSubstitution:CreateForAlphaRename(TypeSubstitution,ImmutableArray`1):TypeSubstitution
         -23 (-3.65% of base) : Microsoft.CodeAnalysis.CSharp.dasm - InitializerSemanticModel:BindEqualsValue(Binder,EqualsValueClauseSyntax,DiagnosticBag):BoundNode:this
        -220 (-3.29% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:BindObjectCreationExpression(VisualBasicSyntaxNode,ArgumentListSyntax,TypeSymbol,VisualBasicSyntaxNode,ImmutableArray`1,ImmutableArray`1,BoundObjectInitializerExpressionBase,DiagnosticBag,VisualBasicSyntaxNode):BoundExpression:this
         -16 (-3.23% of base) : Microsoft.CodeAnalysis.CSharp.dasm - MetadataDecoder:IsOrClosedOverATypeFromAssemblies(TypeSymbol,ImmutableArray`1):bool
         -49 (-3.22% of base) : System.DirectoryServices.AccountManagement.dasm - StoreCtx:BuildFilterSet(Principal,ref,QbeFilterDescription):this
          -7 (-3.03% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - PEModuleBuilder:GetExportedTypes(NamespaceOrTypeSymbol,ArrayBuilder`1):this
         -17 (-3.02% of base) : Microsoft.CodeAnalysis.VisualBasic.dasm - Binder:VerifyTypeCharacterConsistency(SimpleNameSyntax,TypeSymbol,DiagnosticBag):this
          -3 (-2.94% of base) : Microsoft.CodeAnalysis.CSharp.dasm - LocalBinderFactory:VisitGlobalStatement(GlobalStatementSyntax):this

409 total methods with Code Size differences (237 improved, 172 regressed), 268149 unchanged.
```

### CODE REVIEW

**Comment by EgorBo** [comment]
@AndyAyersMS PTAL

**Comment by AndyAyersMS** [comment]
`optSetBlockWeights` is also somewhat ad-hoc. Can you look at what happens if we have PGO data? 

With PGO we still won't know the bias of the internal branches currently, so the conditional block weights will still be guesses. For the castclass cases we should guess the cast will succeed. For isinst it's not as clear what the right guesses are.

We could consider adding probes to capture the outcome of casts so we could properly weight these blocks. I am not sure if it is worth the trouble.

**Comment by EgorBo** [comment]
@AndyAyersMS 
Right, currently we don't instrument internal blocks produced by casts for two reasons:
1) Casts are expanded only when optimizations are enabled
2) We emit counters before we expand casts (qmarks) -- it happens somewhere during global morph

> For the castclass cases we should guess the cast will succeed

We usually expand them like this:
1) If the target class is exact (final):
```c
if (obj == null)
    return null;
else if (obj->pMT == targetHandle)
    return obj;
else
    return CORINFO_HELP_CHKCASTCLASS_SPECIAL(); // for throw InvalidCastException
```
the last block is cold, the first two are 50/50 in the general case.

2) if the target is not exact:
```c
if (obj == null)
    return null;
else
    return CORINFO_HELP_CHKCASTCLASS_SPECIAL();
```
In this case both are 50/50

**Comment by EgorBo** [comment]
Ah I see now.

Looks like we can guess cast results by real successors. e.g.

image

^ green are internal blocks, blue - real ones with IBC data.

Currently for BB2 we set 0.5 weight while we can guess by BB3 that BB1 always goes to BB2.

**Comment by EgorBo** [comment]
New codegen for the above ^:

image


in case if the input is null or non-string we can't really guess the exact reason what really happened so we assume it's either null or non-string.

**Comment by EgorBo** [comment]
Case 1: input is always a valid string:

image

**Comment by EgorBo** [comment]
Case 2: input is null or invalid type (we can't really differ these two cases by successors):

image

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -17681,6 +17681,9 @@ void Compiler::fgExpandQmarkForCastInstOf(BasicBlock* block, Statement* stmt)
 
     assert(qmark->gtFlags & GTF_QMARK_CAST_INSTOF);
 
+    float currBbWeight = block->bbWeight;
```
Comment (BruceForstall) [inline src/coreclr/jit/morph.cpp:17684]: float => weight_t

File: src/coreclr/jit/morph.cpp [hunk 2]
```diff
@@ -17766,11 +17769,18 @@ void Compiler::fgExpandQmarkForCastInstOf(BasicBlock* block, Statement* stmt)
     cond1Block->bbJumpDest = remainderBlock;
     cond2Block->bbJumpDest = remainderBlock;
 
-    // Set the weights; some are guesses.
+    // Currently, we don't instrument internal blocks, so the only way we can set weights to these blocks
+    // is to analyze successors and take a guess.
+    float cond2BlockWeight = 0;
```
Comment (BruceForstall) [inline src/coreclr/jit/morph.cpp:17774]: float => weight_t
0 => BB_ZERO_WEIGHT

File: src/coreclr/jit/morph.cpp [hunk 3]
```diff
@@ -17766,11 +17769,18 @@ void Compiler::fgExpandQmarkForCastInstOf(BasicBlock* block, Statement* stmt)
...
+    if (currBbWeight > 0)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/morph.cpp:17775]: This needs a comment explaining why the computation makes sense.

I don't think values of 100 and 50 should enter in directly (the scaling should be based entirely on the weights from neighboring blocks and perhaps some ad-hoc guesses which you should call out explicitly).

File: src/coreclr/jit/morph.cpp (hunk 2 above)
Comment (AndyAyersMS) [inline src/coreclr/jit/morph.cpp:17774]: This is a relative probability so perhaps name it like so: (?)
```suggestion
    BasicBlock::weight_t cond2BlockLikelihood= 0;
```

File: src/coreclr/jit/morph.cpp (hunk 1 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/morph.cpp:17684]: Also please use `const` as it makes it easier to see this isn't going to be messed with below.

File: src/coreclr/jit/morph.cpp [hunk 4]
```diff
@@ -17681,6 +17681,9 @@ void Compiler::fgExpandQmarkForCastInstOf(BasicBlock* block, Statement* stmt)
 
     assert(qmark->gtFlags & GTF_QMARK_CAST_INSTOF);
 
+    float currBbWeight = block->bbWeight;
+    float nextBbWeight = (block->bbNext != nullptr) ? block->bbNext->bbWeight : currBbWeight;
```
Comment (AndyAyersMS) [inline src/coreclr/jit/morph.cpp:17685]: Ditto here...

File: src/coreclr/jit/morph.cpp (hunk 3 above)
Comment (EgorBo) [reply src/coreclr/jit/morph.cpp]: I've added some comments around (covers 50 value)

**Comment by EgorBo** [comment]
Will get back to it after https://github.com/dotnet/runtime/pull/50490

**Pull request closed without merging**